### PR TITLE
Register Purchase of Video

### PR DESCRIPTION
**Pull request by aeftimia**
video purchases are recorded redundantly in the VideoStore in two maps:
* address of buyer => sha3 hashes of videos purchased
* sha3 hash of video purchased => address of buyers

tests for both storage types are included in VideoStore.js

### CODE REVIEW

**Comment by aeftimia** [comment]
This should take care of #7 and #3 though sales are registered redundantly in the `VideoStore` contract.

File: contracts/paratii/VideoStore.sol [hunk 1]
```diff
@@ -51,10 +54,35 @@ contract VideoStore is Ownable, Debug {
...
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function userOwns(address user, string videoId) returns(bool) {
+        bytes32[] video_hashes = user_purchases[user];
+        bytes32 videoId_hash = sha3(videoId);
+        for (uint i=0;i<video_hashes.length;++i) {
```
Comment (jellegerbrandy) [inline contracts/paratii/VideoStore.sol]: You need to be a aware of the fact that this is a dangerous pattern in solidity. The `video_hashes` can be of arbitrary legnth, and if it is too long, the transaction will run out of gas and the userOwns function will throw an error. 

There are different patterns to work aroudn this. But the first question is: why is `video_hashes` an array and not a mapping? (If it were a mapping, you would not need the `for`-loop here).

File: contracts/paratii/VideoStore.sol [hunk 2]
```diff
@@ -51,10 +54,35 @@ contract VideoStore is Ownable, Debug {
...
+       user_purchases[buyer].push(video_hash);
+       video_sales[video_hash].push(buyer);
        LogBuyVideo(videoId, buyer, price);
        return true;
     }
 
+    function videoPurchased(string videoId, address user) returns(bool) {
+        bytes32 video_hash = sha3(videoId);
+        address[] users = video_sales[video_hash];
+        for (uint i=0;i<users.length;++i) {
```
Comment (jellegerbrandy) [inline contracts/paratii/VideoStore.sol]: This is a bug: see the remark below about gas issues.

File: contracts/paratii/VideoStore.sol [hunk 3]
```diff
@@ -19,6 +19,9 @@ contract VideoStore is Ownable, Debug {
 
     ParatiiRegistry public paratiiRegistry;
 
+    mapping (address => bytes32[]) public user_purchases;
+    mapping (bytes32 => address[]) public video_sales;
```
Comment (jellegerbrandy) [inline contracts/paratii/VideoStore.sol]: Ok, so duplicating data does resolve the problem of "symmety" of the data, but it is expensive to duplicate data, and it also is a potential source of bugs (accidentally update user_purches without updating video_sales, and you have inconsistent data). 

If we only need to _check_ wheter a user purches a video or not, we can probably do with one single data structure. (i.e. drop `video_sales`). The question is then whether there are use cases where we need to do more than just checking if the purchase has occurred or not.

Comment (aeftimia) [reply contracts/paratii/VideoStore.sol]: `UserRegistry` stores videos liked or disliked by the address of the user and currently has an unused `isAcquired` variable in the video struct. I'm getting rid of the `user_purchases` and integrating the `isAcquired` variable into the registration.

**Comment by aeftimia** [comment]
Replaced `user_purchases` with the `isAcquired` field within `VideoInfo` structs in the `UserRegistry`.
Also stopped storing sha3 hashes as `byte32` objects.

**NOTE:** We may not want to automatically acquire videos when users like or
dislike them. We may, in fact, want to check that user's *can't* like or
dislike videos they have not acquired with an appropriate modifier on the `likeVideo` and `dislikeVideo` functions.

**Comment by aeftimia** [comment]
* Register purchase of videos within the `VideoStore` using a `videoSales` mapping from `videoId`s to `address`es of users that purchased those videos.
*  Redundant registration within the `UserRegistry` using the previously unused `isAcquired` flag in within the `UserRegistry.VideoInfo` struct.

**Pull request merged**